File: packages/web/convex/schedules.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";

export const list = query({
  args: {},
  handler: async (ctx) => {
    return await ctx.db.query("scheduledPipelines").collect();
  },
});

export const listByProject = query({
  args: { projectSlug: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("scheduledPipelines")
      .withIndex("by_project", (q) => q.eq("projectSlug", args.projectSlug))
      .collect();
  },
});

export const listActive = query({
  args: {},
  handler: async (ctx) => {
    return await ctx.db
      .query("scheduledPipelines")
      .withIndex("by_status", (q) => q.eq("status", "active"))
      .filter((q) => q.eq(q.field("enabled"), true))
      .collect();
  },
});

export const get = query({
  args: { id: v.id("scheduledPipelines") },
  handler: async (ctx, args) => {
    return await ctx.db.get(args.id);
  },
});

export const create = mutation({
  args: {
    projectSlug: v.string(),
    pipelineSlug: v.string(),
    cron: v.optional(v.string()),
    frequency: v.optional(v.string()),
    windowEndsAt: v.optional(v.number()),
    enabled: v.boolean(),
    status: v.string(),
    nextRunAt: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const now = Date.now();
    return await ctx.db.insert("scheduledPipelines", {
      ...args,
      createdAt: now,
      updatedAt: now,
    });
  },
});

export const update = mutation({
  args: {
    id: v.id("scheduledPipelines"),
    cron: v.optional(v.string()),
    frequency: v.optional(v.string()),
    windowEndsAt: v.optional(v.number()),
    enabled: v.optional(v.boolean()),
    status: v.optional(v.string()),
    lastRunAt: v.optional(v.number()),
    lastJobId: v.optional(v.string()),
    nextRunAt: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const { id, ...updates } = args;
    const now = Date.now();

    // Remove undefined values
    const cleanUpdates = Object.fromEntries(
      Object.entries(updates).filter(([_, v]) => v !== undefined)
    );

    await ctx.db.patch(id, {
      ...cleanUpdates,
      updatedAt: now,
    });
    return await ctx.db.get(id);
  },
});

export const pause = mutation({
  args: { id: v.id("scheduledPipelines") },
  handler: async (ctx, args) => {
    await ctx.db.patch(args.id, {
      enabled: false,
      status: "paused",
      updatedAt: Date.now(),
    });
    return await ctx.db.get(args.id);
  },
});

export const resume = mutation({
  args: { id: v.id("scheduledPipelines") },
  handler: async (ctx, args) => {
    await ctx.db.patch(args.id, {
      enabled: true,
      status: "active",
      updatedAt: Date.now(),
    });
    return await ctx.db.get(args.id);
  },
});

export const remove = mutation({
  args: { id: v.id("scheduledPipelines") },
  handler: async (ctx, args) => {
    await ctx.db.delete(args.id);
  },
});

File: packages/api/app/routers/schedules.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from app.services.convex_client import convex
from app.services.scheduler_service import parse_frequency, parse_window
import time

router = APIRouter(prefix="/schedules", tags=["schedules"])

class CreateScheduleRequest(BaseModel):
    project_slug: str
    pipeline_slug: str
    frequency: str          # "1h", "daily", "0 * * * *" (cron)
    window: str | None = None  # "7d", "1w"
    enabled: bool = True

class UpdateScheduleRequest(BaseModel):
    cron: str | None = None
    frequency: str | None = None
    window: str | None = None
    enabled: bool | None = None

@router.get("/")
async def list_schedules(project_slug: str | None = None):
    if project_slug:
        return await convex.query("schedules:listByProject", {"projectSlug": project_slug})
    return await convex.query("schedules:list", {})

@router.get("/{id}")
async def get_schedule(id: str):
    sched = await convex.query("schedules:get", {"id": id})
    if not sched:
        raise HTTPException(status_code=404, detail="Schedule not found")
    return sched

@router.post("/")
async def create_schedule(req: CreateScheduleRequest):
    try:
        # Check if frequency is already a cron expression
        if " " in req.frequency:
            cron = req.frequency
        else:
            cron = parse_frequency(req.frequency)

        window_ends_at = None
        if req.window:
            window_ends_at = parse_window(req.window)

        # Initial next_run computation could be done here,
        # but for simplicity let's just trigger it or rely on scheduler fallback
        from croniter import croniter
        from datetime import datetime
        itr = croniter(cron, datetime.now())
        next_dt = itr.get_next(datetime)
        next_ms = int(next_dt.timestamp() * 1000)

        payload = {
            "projectSlug": req.project_slug,
            "pipelineSlug": req.pipeline_slug,
            "cron": cron,
            "frequency": req.frequency,
            "enabled": req.enabled,
            "status": "active" if req.enabled else "paused",
            "nextRunAt": next_ms
        }
        if window_ends_at is not None:
            payload["windowEndsAt"] = window_ends_at

        sched_id = await convex.mutation("schedules:create", payload)
        return {"id": sched_id}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.put("/{id}")
async def update_schedule(id: str, req: UpdateScheduleRequest):
    try:
        updates = {"id": id}
        if req.frequency is not None:
            if " " in req.frequency:
                updates["cron"] = req.frequency
            else:
                updates["cron"] = parse_frequency(req.frequency)
            updates["frequency"] = req.frequency

        if req.window is not None:
            updates["windowEndsAt"] = parse_window(req.window)

        if req.enabled is not None:
            updates["enabled"] = req.enabled
            updates["status"] = "active" if req.enabled else "paused"

        return await convex.mutation("schedules:update", updates)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.delete("/{id}")
async def remove_schedule(id: str):
    await convex.mutation("schedules:remove", {"id": id})
    return {"status": "ok"}

@router.post("/{id}/pause")
async def pause_schedule(id: str):
    return await convex.mutation("schedules:pause", {"id": id})

@router.post("/{id}/resume")
async def resume_schedule(id: str):
    return await convex.mutation("schedules:resume", {"id": id})
